src/iuse.cpp

```diff
@@ -3042,6 +3042,65 @@ int iuse::jackhammer( player *p, item *it, bool, const tripoint & )
     return it->type->charges_to_use();
 }
 
+int iuse::drill( player *p, item *it, bool, const tripoint & )
+{
+    // use has_enough_charges to check for UPS availability
+    // p is assumed to exist for iuse cases
+    if( !p->has_enough_charges( *it, false ) ) {
+        return 0;
+    }
+
+    if( p->is_underwater() ) {
+        p->add_msg_if_player( m_info, _( "You can't do that while underwater." ) );
+        return 0;
+    }
+
+    const cata::optional<tripoint> pnt_ = choose_adjacent( _( "Drill where?" ) );
+    if( !pnt_ ) {
+        return 0;
+    }
+    const tripoint pnt = *pnt_;
+
+    if( pnt == p->pos() ) {
+        p->add_msg_if_player( _( "Don't you have enough holes already?" ) );
+        p->add_msg_if_player( _( "You don't need to tear yourself a new one." ) );
+        return 0;
+    }
+    int charges = it->type->charges_to_use() * 5;
+    const ter_id type = g->m.ter( pnt );
+    const furn_id furn = g->m.furn( pnt );
+    time_duration duration = 15_minutes;
+    if( type == t_chaingate_l || type == t_door_locked || type == t_door_locked_alarm ||
+        type == t_door_locked_interior || type == t_door_locked_peep || type == t_door_metal_pickable ||
+        type == t_door_bar_locked || type == t_door_metal_locked ) {
+        duration = 15_minutes;
+    } else if( furn == f_safe_l || furn == f_gunsafe_ml || furn == f_gunsafe_el ||
+               furn == f_gunsafe_mj ) {
+        if( p->get_skill_level( skill_mechanics ) < 2 ) {
+            p->add_msg_if_player( m_bad,
+                                  _( "Your expertise in mechanics is insufficient to properly drill the locking mechanism." ) );
+            return 0;
+        }
+        duration = std::max( 150_minutes - 20_minutes * ( p->get_skill_level( skill_mechanics ) - 3 ) -
+                             10_minutes * ( p->get_per() - 8 ), 30_minutes );
```

Comment:
Low perception (especially real low, such as 1 or 2) could make the process real long, up to three and a half hours. Just an observation.

Reply:
I've used existing math used in stethoscope action, and it seemed proper. Drilling thick metal is not a quick job when you're blind as a mole (low per).